File: adi_shipment/adi_shipment/page/shiprocket_dashboard/shiprocket_dashboard.py
import frappe
import requests
import json
import re
from adi_shipment.api.shiprocket import get_token

@frappe.whitelist()
def get_dashboard_data(search_term=None, tab_view="All", from_date=None, to_date=None):
    token = get_token()
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {token}"
    }
    
    data = {
        "orders": [],
        "stats": {
            "total": 0,
            "new": 0,
            "pickup_scheduled": 0,
            "in_transit": 0,
            "delivered": 0,
            "cancelled": 0,
            "rto": 0
        }
    }

    try:
        # Shiprocket API Params
        params = {
            "per_page": 100, 
            "sort": "DESC"
        }
        
        if search_term:
            params["search"] = search_term
        
        if from_date:
            params["from"] = from_date
        if to_date:
            params["to"] = to_date

        url = "https://apiv2.shiprocket.in/v1/external/orders"
        response = requests.get(url, headers=headers, params=params)
        
        if response.status_code == 200:
            res_data = response.json()
            all_orders = res_data.get("data", [])
            
            filtered_orders = []
            
            for order in all_orders:
                status = order.get("status", "").upper()
                
                # --- Stats Calculation ---
                data["stats"]["total"] += 1
                
                # Mapping logic based on Shiprocket status codes/text
                if status in ["NEW", "PROCESSING", "INVOICED", "AWB ASSIGNED"]: 
                    data["stats"]["new"] += 1
                elif status in ["READY TO SHIP", "MANIFEST GENERATED"]:
                    data["stats"]["new"] += 1 # Grouping Ready to Ship with New for this card, or we could add separate
                elif "PICKUP" in status and "SCHEDULED" in status:
                    data["stats"]["pickup_scheduled"] += 1
                elif "TRANSIT" in status or status in ["SHIPPED", "OUT FOR DELIVERY", "REACHED AT DESTINATION HUB"]:
                    data["stats"]["in_transit"] += 1
                elif status == "DELIVERED":
                    data["stats"]["delivered"] += 1
                elif status in ["CANCELED", "CANCELLED"]:
                    data["stats"]["cancelled"] += 1
                elif "RTO" in status:
                    data["stats"]["rto"] += 1

                if not status:
                     continue
                
                # --- Pre-process Data (Name, Phone) ---
                c_name = order.get("customer_name", "")
                if "-" in c_name:
                    parts = c_name.split("-")
                    if len(parts) == 2 and parts[0].strip().lower() == parts[1].strip().lower():
                        c_name = parts[0].strip()
                order["customer_name"] = c_name
                
                # Fetch mobile number from shipment doc
                if order.get('channel_order_id', '').startswith('SHIP-'):
                    try:
                        shipment_name = order.get('channel_order_id')
                        delivery_contact = frappe.db.get_value('Shipment', shipment_name, 'delivery_contact')
                        if delivery_contact:
                            phones = re.findall(r'\b\d{10}\b', str(delivery_contact))
                            if phones:
                                order['customer_phone'] = phones[-1]
                    except Exception:
                        pass
                
                # Extract Shipment Details (AWB, Courier) early for search
                s_awb = ""
                s_courier = ""
                if order.get("shipments"):
                     s_awb = str(order.get("shipments")[0].get("awb", "")).lower()
                     s_courier = str(order.get("shipments")[0].get("courier", "")).lower()
                
                # --- Search Filter ---
                if search_term:
                    term = search_term.lower()
                    # Fields to search in
                    s_id = str(order.get("id", "")).lower()
                    c_id = str(order.get("channel_order_id", "")).lower()
                    c_name_lower = str(order.get("customer_name", "")).lower()
                    c_email = str(order.get("customer_email", "")).lower()
                    c_phone = str(order.get("customer_phone", "")).lower()

                    if (term not in s_id and 
                        term not in c_id and 
                        term not in c_name_lower and 
                        term not in c_email and 
                        term not in c_phone and
                        term not in s_awb and
                        term not in s_courier):
                        continue

                # --- Tab Filtering ---
                include_order = False
                
                if tab_view == "All":
                    include_order = True
                
                elif tab_view == "New":
                    # Pending orders
                    if status in ["NEW", "PROCESSING"]:
                        include_order = True
                        
                elif tab_view == "Ready To Ship":
                    # Invoice Generated / AWB Assigned but not pickup scheduled yet
                    # Merged NEW/PROCESSING here as per user request to remove NEW tab
                    if status in ["NEW", "PROCESSING", "INVOICED", "AWB ASSIGNED", "READY TO SHIP", "MANIFEST GENERATED"]:
                        include_order = True
                        
                elif tab_view == "Pickups":
                    # Pickup scheduled or queued
                    if "PICKUP" in status and ("SCHEDULED" in status or "QUEUED" in status or "RESCHEDULED" in status):
                        include_order = True
                        
                elif tab_view == "In Transit":
                    if "TRANSIT" in status or status in ["SHIPPED", "OUT FOR DELIVERY", "REACHED AT DESTINATION HUB"]:
                        include_order = True
                        
                elif tab_view == "Delivered":
                    if status == "DELIVERED":
                        include_order = True
                        
                elif tab_view == "RTO":
                    if "RTO" in status:
                        include_order = True
                        
                elif tab_view == "Cancelled":
                    if status in ["CANCELED", "CANCELLED"]:
                        include_order = True

                if include_order:
                    # Enrich order object for UI
                    
                    # Payment extraction (sometimes it's in payment_method or others)
                    # No change needed if already working, just standardizing
                    
                    # Extract courier and awb
                    s_details = order.get("shipments")
                    order["display_courier"] = s_details[0].get("courier") if s_details else ""
                    order["display_awb"] = s_details[0].get("awb") if s_details else ""
                    
                    filtered_orders.append(order)
            
            data["orders"] = filtered_orders
                    
        else:
             frappe.log_error(f"Shiprocket Dashboard Error: {response.text}")

    except Exception as e:
        frappe.log_error(f"Shiprocket Dashboard Exception: {str(e)}")
        
    return data

@frappe.whitelist()
def print_shiprocket_artifact(shipment_id, type="label"):
    """
    Generate Label or Manifest directly from Shipment ID (Shiprocket ID).
    """
    token = get_token()
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {token}"
    }

    url = ""
    payload = {"shipment_id": [shipment_id]}

    if type == "label":
        url = "https://apiv2.shiprocket.in/v1/external/courier/generate/label"
    elif type == "manifest":
        url = "https://apiv2.shiprocket.in/v1/external/manifests/generate"
    else:
        frappe.throw("Invalid Print Type")

    try:
        response = requests.post(url, json=payload, headers=headers)
        data = response.json()
        
        if type == "label":
             if data.get("label_url"):
                 return {"url": data.get("label_url")}
             elif data.get("label_created") == 1 and data.get("response_url"):
                 return {"url": data.get("response_url")}
             else:
                 frappe.throw(f"Could not generate label: {data.get('message', 'Unknown Error')}")
        
        elif type == "manifest":
             if data.get("manifest_url"):
                  return {"url": data.get("manifest_url")}
             else:
                  frappe.throw(f"Could not generate manifest: {data.get('message', 'Unknown Error')}")

    except Exception as e:
        frappe.log_error(f"Print Artifact Error: {str(e)}")
        frappe.throw(str(e))

File: adi_shipment/adi_shipment/page/shiprocket_dashboard/shiprocket_dashboard.js
frappe.pages['shiprocket_dashboard'].on_page_load = function (wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Shiprocket Dashboard',
        single_column: true
    });

    wrapper.page = page;

    // -- Top Filters --
    let $filter_row = $(`<div class="row filter-row" style="margin-bottom: 20px;"></div>`).appendTo(page.main);

    // 1. Search Filter
    $filter_row.append(`
        <div class="col-md-5">
            <div class="form-group">
                <input type="text" class="form-control input-sm input-search" placeholder="Search Order ID, AWB, Mobile..." style="border-radius: 20px; padding-left: 15px;">
            </div>
        </div>
    `);

    // 2. Date Filter (Last 30 Days default, but we use a date range picker logic if possible, or just two fields)
    // Using standard frappe date fields is easier
    $filter_row.append(`
        <div class="col-md-3">
            <input type="date" class="form-control input-sm input-date-from" placeholder="From Date" style="border-radius: 20px;">
        </div>
        <div class="col-md-3">
            <input type="date" class="form-control input-sm input-date-to" placeholder="To Date" style="border-radius: 20px;">
        </div>
    `);

    // 3. Refresh Button (Icon)
    $filter_row.append(`
        <div class="col-md-1 text-right">
             <button class="btn btn-default btn-refresh btn-sm" style="border-radius: 50%; width: 32px; height: 32px; padding: 0;"><i class="fa fa-refresh"></i></button>
        </div>
    `);

    // -- Tabs --
    let tabs = ["All", "New", "Ready To Ship", "Pickups", "In Transit", "Delivered", "RTO", "Cancelled"];
    let $tab_nav = $(`<ul class="nav nav-tabs" style="margin-bottom: 20px; border-bottom: 2px solid #ddd;"></ul>`).appendTo(page.main);

    tabs.forEach((tab, index) => {
        let active = index === 0 ? "active" : "";
        $tab_nav.append(`
            <li class="${active}">
                <a href="#" class="tab-link" data-tab="${tab}" style="border:none; background:transparent; font-weight: 500; font-size: 13px;">${tab}</a>
            </li>
        `);
    });

    // -- Content Container --
    let $content = $(`<div class="dashboard-content"></div>`).appendTo(page.main);

    // Bind Events
    wrapper.active_tab = "All";

    $tab_nav.on('click', '.tab-link', function (e) {
        e.preventDefault();
        $tab_nav.find('li').removeClass('active');
        $(this).parent().addClass('active');
        wrapper.active_tab = $(this).data('tab');
        render_dashboard(wrapper);
    });

    $filter_row.find('.input-search').on('change', function () {
        wrapper.search_term = $(this).val();
        render_dashboard(wrapper);
    });

    $filter_row.find('.input-date-from').on('change', function () {
        wrapper.from_date = $(this).val();
        render_dashboard(wrapper);
    });

    $filter_row.find('.input-date-to').on('change', function () {
        wrapper.to_date = $(this).val();
        render_dashboard(wrapper);
    });

    $filter_row.find('.btn-refresh').on('click', function () {
        render_dashboard(wrapper);
    });

    // Custom CSS
    add_custom_css();

    render_dashboard(wrapper);
}

function render_dashboard(wrapper) {
    let $container = wrapper.page.main.find('.dashboard-content');
    $container.empty();

    // Loading State
    $container.append(`
        <div class="flex justify-center align-center" style="height: 300px;">
            <div class="text-center">
                <div class="spinner-border text-primary" role="status"></div>
                <p class="text-muted mt-2">Fetching Data...</p>
            </div>
        </div>
    `);

    frappe.call({
        method: "adi_shipment.adi_shipment.page.shiprocket_dashboard.shiprocket_dashboard.get_dashboard_data",
        args: {
            search_term: wrapper.search_term || '',
            tab_view: wrapper.active_tab || 'All',
            from_date: wrapper.from_date,
            to_date: wrapper.to_date
        },
        callback: function (r) {
            $container.empty();

            if (r.message) {
                let data = r.message;
                let stats = data.stats;
                let orders = data.orders || [];

                // 1. Stats Cards
                let stats_html = `
                    <div class="row overview-stats-row">
                        ${get_stat_card("Total Orders", stats.total, "blue", "fa fa-shopping-cart")}
                        ${get_stat_card("New / Processing", stats.new, "orange", "fa fa-clock-o")}
                        ${get_stat_card("Pickup Scheduled", stats.pickup_scheduled, "cyan", "fa fa-truck")}
                        ${get_stat_card("In Transit", stats.in_transit, "purple", "fa fa-plane")}
                        ${get_stat_card("Delivered", stats.delivered, "green", "fa fa-check-circle")}
                        ${get_stat_card("Cancelled", stats.cancelled, "red", "fa fa-ban")}
                    </div>
                `;

                // 2. Orders List
                let table_html = `
                    <div class="frappe-card dashboard-list">
                        <div class="frappe-card-head" style="background:#f9f9f9; padding:10px 15px;">
                            <span class="text-muted">Filtered Orders (${orders.length})</span>
                        </div>
                        <div class="frappe-card-body table-responsive">
                            <table class="table table-hover" style="font-size: 13px;">
                                <thead style="background-color: #f1f1f1; color: #555;">
                                    <tr>
                                        <th>Order Details</th>
                                        <th>Customer</th>
                                        <th>Status</th>
                                        <th>Payment</th>
                                        <th>Courier / AWB</th>
                                        <th class="text-right">Action</th>
                                    </tr>
                                </thead>
                                <tbody>
                                    ${orders.length > 0 ? orders.map(order => get_order_row(order)).join('') : '<tr><td colspan="6" class="text-center text-muted p-5">No orders found.</td></tr>'}
                                </tbody>
                            </table>
                        </div>
                    </div>
                `;

                $container.append(`<div class="dashboard-container">${stats_html} ${table_html}</div>`);
            }
        }
    });
}

function get_stat_card(label, value, color, icon) {
    return `
        <div class="col-md-2 col-sm-4 col-xs-6 mb-3">
            <div class="stat-card" style="border-top: 3px solid ${get_color_hex(color)};">
                <div class="flex justify-between"> 
                    <div class="stat-details">
                        <h3 class="text-${color}" style="margin:0; font-weight:700;">${value}</h3>
                        <span class="text-muted" style="font-size:11px; text-transform:uppercase;">${label}</span>
                    </div>
                    <div class="stat-icon text-${color}">
                        <i class="${icon}"></i>
                    </div>
                </div>
            </div>
        </div>
    `;
}

function get_color_hex(color) {
    const colors = {
        'blue': '#4da1ff',
        'orange': '#ffa00a',
        'cyan': '#00d2d3',
        'purple': '#5f27cd',
        'green': '#2ecc71',
        'red': '#ff6b6b'
    };
    return colors[color] || '#ddd';
}

function get_order_row(order) {
    let status_color = get_status_color(order.status);
    let shipment_id = order.shipments && order.shipments.length > 0 ? order.shipments[0].id : null;
    let awb_code = order.display_awb || (order.shipments && order.shipments.length > 0 ? order.shipments[0].awb : null);
    let tracking_url = awb_code ? `https://shiprocket.co/tracking/${awb_code}` : '#';
    let courier_name = order.display_courier || '';

    // Action Buttons Logic
    let actions = '';

    // Status based actions
    let status = (order.status || '').toUpperCase();

    if (status === 'NEW' || status === 'PROCESSING') {
        // Likely needs scheduling or label
        if (shipment_id) {
            actions += `<button class="btn btn-xs btn-primary-soft mr-1" onclick="print_shiprocket_doc('${shipment_id}', 'label')">Label</button>`;
        }
    } else if (status === 'PICKUP SCHEDULED' || status === 'PICKUP QUEUED') {
        if (shipment_id) {
            actions += `<button class="btn btn-xs btn-default mr-1" onclick="print_shiprocket_doc('${shipment_id}', 'manifest')">Manifest</button>`;
            actions += `<button class="btn btn-xs btn-default mr-1" onclick="print_shiprocket_doc('${shipment_id}', 'label')">Label</button>`;
        }
    } else if (['SHIPPED', 'IN TRANSIT', 'OUT FOR DELIVERY'].includes(status)) {
        if (awb_code) {
            actions += `<a href="${tracking_url}" target="_blank" class="btn btn-xs btn-info-soft mr-1"><i class="fa fa-map-marker"></i> Track</a>`;
        }
    } else if (status === 'DELIVERED') {
        actions += `<span class="text-muted text-xs">Completed</span>`;
    }

    // Default Fallback if no specific action matched but shipment exists
    if (actions === '' && shipment_id) {
        actions += `<button class="btn btn-xs btn-default" onclick="print_shiprocket_doc('${shipment_id}', 'label')"><i class="fa fa-print"></i></button>`;
    }

    return `
        <tr>
            <td>
                <div class="font-weight-bold text-primary" style="cursor:pointer;" onclick="window.open('https://app.shiprocket.in/orders/processing?search=${order.id}', '_blank')">
                    ${order.channel_order_id || order.id}
                </div>
                <div class="text-muted text-xs">${order.created_at}</div>
            </td>
            <td>
                <div class="font-weight-bold" style="font-size:12px;">${order.customer_name}</div>
                <div class="text-muted text-xs">${order.customer_email || ''}</div>
                <div class="text-muted text-xs">${order.customer_phone || ''}</div>
            </td>
            <td>
                <span class="indicator-pill ${status_color}">${order.status}</span>
            </td>
            <td>
                <div><span class="badge badge-light">${order.payment_method}</span></div>
                <div class="font-weight-bold mt-1">₹ ${order.total}</div>
            </td>
             <td>
                <div class="font-weight-bold text-dark" style="font-size:12px;">${courier_name}</div>
                <div class="text-muted text-xs">${awb_code ? `AWB: ${awb_code}` : '-'}</div>
            </td>
            <td class="text-right">
                <div class="btn-group">
                    ${actions}
                </div>
            </td>
        </tr>
    `;
}

// Global function to be called from onclick
window.print_shiprocket_doc = function (shipment_id, type) {
    if (!shipment_id) return;

    let method = type === 'label' ? 'adi_shipment.api.shiprocket.generate_label' : 'adi_shipment.api.shiprocket.generate_manifest_by_id';

    // We need a wrapper because the existing API takes 'shipment_name' (DocName) not shipment_id directly
    // Wait, the existing API takes 'shipment_name' (Frappe Doc).
    // The dashboard lists Shiprocket orders which might NOT be linked to Frappe "Shipment" docs perfectly 1:1 in all cases or we want to use direct API.
    // However, to reuse the Code, we assume we have a wrapper or we call a new python method.
    // Let's create a new python method in dashboard controller to handle raw IDs.

    frappe.call({
        method: "adi_shipment.adi_shipment.page.shiprocket_dashboard.shiprocket_dashboard.print_shiprocket_artifact",
        args: {
            shipment_id: shipment_id,
            type: type
        },
        freeze: true,
        callback: function (r) {
            if (r.message && r.message.url) {
                window.open(r.message.url, '_blank');
            }
        }
    });
}

function get_status_color(status) {
    status = (status || '').toUpperCase();
    if (['NEW', 'PROCESSING', 'INVOICED', 'AWB ASSIGNED'].includes(status)) return 'orange';
    if (['DELIVERED', 'COMPLETED'].includes(status)) return 'green';
    if (['CANCELED', 'CANCELLED'].includes(status)) return 'red';
    if (['SHIPPED', 'PICKUP SCHEDULED', 'OUT FOR DELIVERY'].includes(status) || status.includes('TRANSIT')) return 'purple';
    if (status.includes('RTO')) return 'red';
    return 'blue';
}

function add_custom_css() {
    $('head').append(`
        <style>
            /* Layout & Background */
            .layout-main-section {
                 background-color: #f4f6f8;
            }
            .dashboard-container {
                padding: 10px;
            }

            /* Filters */
            .filter-row .form-control {
                background: #fff;
                border: 1px solid #dfe1e5;
                font-size: 13px;
                box-shadow: none;
            }
            
            /* Tabs */
            .nav-tabs {
                border-bottom: 2px solid #e0e0e0;
            }
            .nav-tabs > li > a {
                color: #555;
                padding: 10px 15px;
                margin-right: 5px;
                border-bottom: 2px solid transparent;
            }
            .nav-tabs > li.active > a, 
            .nav-tabs > li.active > a:hover, 
            .nav-tabs > li.active > a:focus {
                color: #5f27cd;
                background-color: transparent;
                border: none;
                border-bottom: 2px solid #5f27cd;
            }
            
            /* Stat Cards */
            .stat-card {
                background: #fff;
                border-radius: 8px;
                padding: 15px;
                box-shadow: 0 1px 2px rgba(0,0,0,0.05);
                height: 100%;
                transition: transform 0.2s;
            }
            .stat-card:hover {
                 transform: translateY(-2px);
                 box-shadow: 0 4px 10px rgba(0,0,0,0.05);
            }
            
            /* Table */
            .dashboard-list {
                background: #fff;
                border-radius: 8px;
                border: 1px solid #e0e0e0;
                box-shadow: 0 1px 2px rgba(0,0,0,0.05);
                overflow: hidden;
            }
            .table-hover tbody tr:hover {
                background-color: #f9faff;
            }
            
            /* Buttons */
            .btn-primary-soft {
                background-color: #e8eaf6;
                color: #3f51b5;
                border: none;
            }
            .btn-primary-soft:hover {
                background-color: #c5cae9;
                color: #303f9f;
            }
             .btn-info-soft {
                background-color: #e1f5fe;
                color: #039be5;
                border: none;
            }
            
            /* Indicators */
            .indicator-pill {
                padding: 3px 8px;
                border-radius: 4px;
                font-size: 10px;
                font-weight: 700;
                text-transform: uppercase;
                letter-spacing: 0.5px;
            }
            .indicator-pill.orange { background: #fff8e1; color: #ff8f00; }
            .indicator-pill.green { background: #e8f5e9; color: #388e3c; }
            .indicator-pill.purple { background: #f3e5f5; color: #8e24aa; }
            .indicator-pill.red { background: #ffebee; color: #d32f2f; }
            .indicator-pill.blue { background: #e3f2fd; color: #1976d2; }
            
        </style>
    `);
}
